### PR TITLE
Hide PRCA behind a build variable flag

### PR DESCRIPTION
**Pull request by bgavrilMS**

### CODE REVIEW

File: Tasks/SonarQubePostTest/SonarQubePostTest.ps1 [hunk 1]
```diff
@@ -4,15 +4,16 @@ import-module "Microsoft.TeamFoundation.DistributedTask.Task.Common"
...
 
-if ( (IsPrBuild) -and ((GetTaskContextVariable "DisableSQAnalysisOnPrBuilds") -eq "true"))
+# During PR builds only an "issues mode" analysis is allowed. The resulting issues are posted as code review comments. 
+# The feature can be toggled by the user and is OFF by default.  
+if (ShouldExitOnPRBuild)
 {
-	Write-Host "DisableSQAnalysisOnPrBuilds is set and this is a PR build - ignoring the analysis tasks"
-	return
+    Write-Host "SonarQube analysis is disabled during builds triggered by pull requests. Set a build variable named 'SQPullRequestBot' to 'true' to have the task post code analysis issues as comments in the PR. More information at http://go.microsoft.com/fwlink/?LinkID=786316"
+    exit
```
Comment (duncanpMS) [inline Tasks/SonarQubePostTest/SonarQubePostTest.ps1]: Why the change from _return_ to _exit_?

File: Tasks/SonarQubePreBuild/SonarQubePreBuild.ps1 [hunk 2]
```diff
@@ -27,30 +27,17 @@ import-module "Microsoft.TeamFoundation.DistributedTask.Task.Internal"
...
 . $PSScriptRoot/Common/SonarQubeHelpers/SonarQubeHelper.ps1
 
-if ( (IsPrBuild) -and ((GetTaskContextVariable "DisableSQAnalysisOnPrBuilds") -eq "true"))
+# During PR builds only an "issues mode" analysis is allowed. The resulting issues are posted as code review comments. 
+# The feature can be toggled by the user and is OFF by default.  
+if (ShouldExitOnPRBuild)
 {
-	Write-Host "DisableSQAnalysisOnPrBuilds is set and this is a PR build - ignoring the analysis tasks"	
-	return
+    Write-Host "SonarQube analysis is disabled during builds triggered by pull requests. Set a build variable named 'SQPullRequestBot' to 'true' to have the task post code analysis issues as comments in the PR. More information at http://go.microsoft.com/fwlink/?LinkID=786316"
```
Comment (duncanpMS) [inline Tasks/SonarQubePreBuild/SonarQubePreBuild.ps1]: Is it necessary to write the same message in both the _begin_ and _end_ tasks - isn't just the _begin_ task enough?

If you think it is necessary to issue the message twice, could you avoid the code duplication by pushing all of the logic into the helper method and have the helper method write the message then exit?
Also, is all of the duplication necessary? If you're using _exit_ couldn't you put all

File: Tasks/SonarQubePreBuild/SonarQubePreBuildImpl.ps1 [hunk 3]
```diff
@@ -1,4 +1,27 @@
...
+
+    $cmdLineArgs = UpdateArgsForPullRequestAnalysis $cmdLineArgs $serviceEndpoint
+    Write-Verbose -Verbose $cmdLineArgs
+
+    $arguments = CreateCommandLineArgs $projectKey $projectName $projectVersion $serviceEndpoint.Url $serviceEndpoint.Authorization.Parameters.UserName $serviceEndpoint.Authorization.Parameters.Password $dbUrl $dbUsername $dbPassword $cmdLineArgs $configFile
+
+    Invoke-BatchScript $bootstrapperPath -Arguments $arguments
+}
+
+#
```
Comment (duncanpMS) [inline Tasks/SonarQubePreBuild/SonarQubePreBuildImpl.ps1]: It doesn't look like any of this code has changed. However, it's not that easy to tell in the web based UI tools. Please help out reviewers by calling out code that just been moved (and if it's been moved and changed then call out the changes).

File: Tests/L0/SonarQubePostTest/DisableAnalysisOnPrBuild.ps1 [hunk 4]
```diff
@@ -4,13 +4,68 @@ param()
...
+        Assert-WasCalled InvokeMSBuildRunnerPostTest
+    }
+    else
+    {
+        Assert-WasCalled Write-Host -ArgumentsEvaluator {$args[0] -like "*SQPullRequestBot*"}
+        Assert-WasCalled InvokeMSBuildRunnerPostTest -Times 0
+    }
+    
+    # Cleanup
+    Unregister-Mock IsPrBuild
```
Comment (duncanpMS) [inline Tests/L0/SonarQubePostTest/DisableAnalysisOnPrBuild.ps1]: Out of interest, how do the _Assert_ methods work? Presumably they don't halt execution otherwise the clean-up code won't be hit if an assertion fails.

**Comment by duncanpMS** [comment]
LGTM

File: Tasks/SonarQubePreBuild/SonarQubePreBuildImpl.ps1 (hunk 3 above)
Comment (bgavrilMS) [inline Tasks/SonarQubePreBuild/SonarQubePreBuildImpl.ps1]: Indeed, sorry about that.

File: Tests/L0/SonarQubePostTest/DisableAnalysisOnPrBuild.ps1 (hunk 4 above)
Comment (bgavrilMS) [inline Tests/L0/SonarQubePostTest/DisableAnalysisOnPrBuild.ps1]: Asserts just throw exceptions so they will halt execution and the cleanup will not happen. However, the cleanup is only relevant inside this file. When another test file is invoked, the mocks are no longer in effect.
